### PR TITLE
Set chapter as module model answer

### LINKED ISSUE #1098: It should be possible to control chapter visibility in the same way as model solution visibility in assignments
Especially the O1 course (Programming 1, @jsorva) regularly uses chapters at the start of modules that provide feedback and commentary on the previous module and its assignments. In practice, the chapter also reveals model solutions for some of the assignments in the past module. The normal deadline of the past module has passed at this point, but there are always students who have personal deadline extensions. It should be possible to set full chapters as model solutions whose visibility can be controlled with the same settings that are available to assignment model solutions. Assignments already support flexible settings for the visibility of model solutions.

Assignment model solutions currently use the `BaseExercise.model_solutions_reveal_rule` field to control their visibility.
https://github.com/apluslms/a-plus/blob/34cc48150212b0ff9ebd55c7c47e97f21b07f706/exercise/exercise_models.py#L520

Reveal rule and exercise reveal state classes:
* https://github.com/apluslms/a-plus/blob/34cc48150212b0ff9ebd55c7c47e97f21b07f706/exercise/reveal_models.py#L12
* https://github.com/apluslms/a-plus/blob/34cc48150212b0ff9ebd55c7c47e97f21b07f706/exercise/reveal_states.py#L35

User docs for the existing reveal rules in the A-plus-rst-tools README:
https://github.com/apluslms/a-plus-rst-tools/blob/40c904b69e16518bf8e01cffac70877b58bbc448/README.md#defining-reveal-rules

### PR DESCRIPTION
**Pull request by Mikael-Lenander**
# Description

**What?**

Modules can be configured a chapter as model answer.  The chapters visibility is controlled through a RevealRule.

**Why?**

Chapters may include prujus to previous modules, especially in O1.

**How?**

Adds model answer and reveal rule fields to CourseModule. A Chapter can check (by reversing the relation) if it is a model solution to a module. Any LearningObject can check if it is a child of this kind of Chapter. Adds a key 'is_revealed' to every LearningObject entry in cache. Updates exercise permission class to check LearningObject's visibility. Adds new reveal state that takes all module's exercises into account.

Fixes #1098

# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)

**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [x] Django unit tests.
- [x] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [x] Manual testing.

Made chapter 2.0 module 1's model answer. Added deadline deviation to module 1 for student so that module 1 and 2 are open simultaneously. Made sure chapter 2.0 is not visible to student, not through toc, results or direct link.

**Did you test the changes in**

- [x] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [x] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [x] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?
- [x] Did you use Python type hinting in all functions that you added or edited? ([type hints](https://docs.python.org/3/library/typing.html) for function parameters and return values)

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

### CODE REVIEW

File: exercise/permissions.py [hunk 1]
```diff
@@ -67,6 +67,10 @@ def is_object_visible(self, request, view, exercise): # pylint: disable=argument
                 self.error_msg(_('EXERCISE_VISIBLITY_ERROR_ONLY_EXTERNAL_USERS'))
                 return False
 
+        if not exercise.can_be_shown_as_module_model_solution(user):
+                self.error_msg(_('MODULE_MODEL_ANSWER_VISIBILITY_PERMISSION_DENIED_MSG'))
+                return False
```
Comment (ihalaij1) [inline exercise/permissions.py:72]: The code in this if-statement is indented twice.

File: exercise/reveal_states.py [hunk 2]
```diff
@@ -106,7 +121,72 @@ def get_latest_deadline(self) -> Optional[datetime.datetime]:
...
+    @overload
+    def __init__(self, module: Dict[str, Any]):
+        ...
+    def __init__(self, module: Union[CourseModule, Dict[str, Any]], student: Optional[User] = None):
+        if isinstance(module, CourseModule):
+            from .cache.points import CachedPoints # pylint: disable=import-outside-toplevel
+            cached_content = CachedContent(module.course_instance)
+            cached_points = CachedPoints(module.course_instance, student, cached_content, True)
+            self.module_id = module.id
+            cached_module,_,_,_ = cached_points.find(module)
```
Comment (ihalaij1) [inline exercise/reveal_states.py:140]: Could you add spaces after the commas?

File: exercise/tests_cache.py [hunk 3]
```diff
@@ -240,3 +241,35 @@ def test_unofficial(self):
...
+        self.exercise.save()
+        self.module0.model_answer = module_chapter
+        self.module0.model_solution_reveal_rule = reveal_rule
+        self.module0.save()
+        c = CachedContent(self.instance)
+        p = CachedPoints(self.instance, self.student, c)
+        entry0,_,_,_ = p.find(self.exercise0)
+        entry,_,_,_ = p.find(self.exercise)
+        entry2,_,_,_ = p.find(self.exercise2)
+        chapter_entry,_,_,_ = p.find(module_chapter)
```
Comment (ihalaij1) [inline exercise/tests_cache.py:271]: Could you add spaces after the commas?

**Review by ihalaij1 — CHANGES_REQUESTED** [review]
Great!
I had just a few comments, it seems to be working.
Nice that you added tests as well!
You should still fix the linter errors, though.

**Review by Mankro — CHANGES_REQUESTED** [review]
You have done great work! There is very little time left now to finish this before the version release. Please, at least fix the trivial parts that I have commented. Bigger performance optimizations could be done in the future if necessary. 

In addition, rebase onto the upstream and squash your commits. I can then merge this.

File: exercise/reveal_states.py [hunk 4]
```diff
@@ -106,7 +121,72 @@ def get_latest_deadline(self) -> Optional[datetime.datetime]:
...
+
+    def get_points(self) -> Optional[int]:
+        points = sum([exercise['points'] for exercise in self.exercises])
+        return points
+
+    def get_max_points(self) -> Optional[int]:
+        return self.module['max_points']
+
+    def get_submissions(self) -> Optional[int]:
+        return sum([min(exercise['submission_count'], _get_max_submissions(exercise)) for exercise in self.exercises])
```
[the remark this answers is not in the document]
Comment (Mikael-Lenander) [reply exercise/reveal_states.py:179]: Generally, if for some reason the submission count is greater than max submission count, we want to use the max submission count.

**Comment by Mikael-Lenander** [comment]
Still fixed linting errors.

File: exercise/reveal_states.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Mankro) [reply exercise/reveal_states.py]: OK! :+1:

**Review by Mankro — APPROVED** [review]
Thanks, I will merge this now.

**Pull request merged**